### PR TITLE
Add photoionization-related rates

### PR DESCRIPTION
**Pull request by chvogl**
In progress ... do not merge

Atomic Data with photoionization data

https://drive.google.com/file/d/1kN419gAzOvrTpboN52GsYiKh3-fG5pXQ/view?usp=sharing

### CODE REVIEW

File: tardis/plasma/properties/continuum_processes.py [hunk 1]
```diff
@@ -89,4 +172,373 @@ def calculate(self, photo_ion_cross_sections, t_electrons,
...
+    outputs = ('coll_exc_coeff',)
+    latex_name = ('c_{lu}',)
+
+    def calculate(self, yg_interp, yg_index, t_electrons, delta_E_yg):
+        yg = yg_interp(t_electrons)
+        k_B = const.k_B.cgs.value
+        boltzmann_factor = np.exp(
+            - delta_E_yg.values[np.newaxis].T / (t_electrons * k_B)
+        )
+        q_ij = 8.629e-6 / np.sqrt(t_electrons) * yg * boltzmann_factor
```
Comment (wkerzendorf) [inline tardis/plasma/properties/continuum_processes.py:418]:
```suggestion
        q_ij = 8.629e-6 / np.sqrt(t_electrons) * yg * boltzmann_factor # see formula A2 in Przybilla, Butler 2004 - Apj 609, 1181
```

File: tardis/plasma/properties/util/macro_atom.pyx [hunk 2]
```diff
@@ -45,17 +46,18 @@ def calculate_transition_probabilities(
             for j in range(transition_probabilities.shape[1]):
                 transition_probabilities[i, j] *= stimulated_emission_factor[line_idx, j] * j_blues[line_idx, j]
 
-    for i in range(block_references.shape[0] - 1):
-        for k in range(transition_probabilities.shape[1]):
-            norm_factor[k] = 0.0
-        for j in range(block_references[i], block_references[i + 1]):
+    if normalize:
```
Comment (andrewfullard) [inline tardis/plasma/properties/util/macro_atom.pyx:49]: I think this should be a wrapper on the function, not just indenting the loop

File: tardis/plasma/base.py [hunk 3]
```diff
@@ -183,6 +183,24 @@ def update(self, **kwargs):
         for module_name in self._resolve_update_list(kwargs.keys()):
             self.plasma_properties_dict[module_name].update()
 
+    def freeze(self, *args):
+        for key in args:
```
Comment (wkerzendorf) [inline tardis/plasma/base.py:187]: documentation

File: tardis/plasma/base.py [hunk 4]
```diff
@@ -183,6 +183,24 @@ def update(self, **kwargs):
...
+            if key not in self.outputs_dict:
+                raise PlasmaMissingModule(
+                    "Trying to freeze property {0}"
+                    " that is unavailable".format(key)
+                )
+            self.outputs_dict[key].frozen = True
+
+    def thaw(self, *args):
+        for key in args:
```
Comment (wkerzendorf) [inline tardis/plasma/base.py:196]: documentation

File: tardis/plasma/properties/atomic.py [hunk 5]
```diff
@@ -298,3 +423,117 @@ def calculate(self, atomic_data):
...
+                "destination_level_idx": destination_idx.values,
+            },
+            index=index,
+        )
+        return yg_data, t_yg, index, delta_E, yg_idx
+
+    @staticmethod
+    def calculate_yg_van_regemorter(
+        atomic_data, t_electrons, continuum_interaction_species
+    ):
```
Comment (wkerzendorf) [inline tardis/plasma/properties/atomic.py:497]: add van regemorter paper here.

File: tardis/plasma/properties/continuum_processes.py [hunk 6]
```diff
@@ -82,20 +254,720 @@ def calculate(
...
+        The adiabatic cooling rate of the electron gas.
+    """
+
+    outputs = ("C_adiabatic",)
+    transition_probabilities_outputs = ("C_adiabatic",)
+    latex_name = (r"C_{\textrm{adiabatic}}",)
+
+    def calculate(self, electron_densities, t_electrons, time_explosion):
+        C_adiabatic = (
+            3.0 * electron_densities * const.k_B.cgs.value * t_electrons
```
Comment (wkerzendorf) [inline tardis/plasma/properties/continuum_processes.py:822]: that might be nice to have it as a `CONSTANT`

File: tardis/plasma/properties/continuum_processes.py [hunk 7]
```diff
@@ -6,9 +6,39 @@
 from numba import prange, njit
 from astropy import constants as const
```
Comment (wkerzendorf) [inline tardis/plasma/properties/continuum_processes.py:7]: this should be `from tardis import constants as const`

File: tardis/plasma/properties/continuum_processes.py [hunk 8]
```diff
@@ -82,20 +254,720 @@ def calculate(
...
+            names=list(yg_idx.columns) + ["transition_type"],
+        )
+        p_exc_cool = p_exc_cool.set_index(exc_cool_index)
+        p_coll = pd.concat(
+            [p_deexc_deac, p_deexc_internal, p_exc_internal, p_exc_cool]
+        )
+        return p_coll
+
+
+class BaseTwoPhotonTransProbs(
```
Comment (wkerzendorf) [inline tardis/plasma/properties/continuum_processes.py:709]: `RawTwoPhoton` ... ?

File: tardis/plasma/base.py (hunk 3 above)
Comment (chvogl) [reply tardis/plasma/base.py:203]: Added a docstring.

File: tardis/plasma/properties/radiative_properties.py [hunk 9]
```diff
@@ -17,8 +20,13 @@
     "TauSobolev",
     "BetaSobolev",
     "TransitionProbabilities",
+    "BaseRadBoundBoundTransProbs",
 ]
 
+C_EINSTEIN = (
+    4.0 * (np.pi * const.e.esu) ** 2 / (const.c.cgs * const.m_e.cgs)
+).value * u.eV.to(u.erg)
```
[the remark this answers is not in the document]
Comment (chvogl) [reply tardis/plasma/properties/radiative_properties.py]: Should be more clear now.

**Pull request merged**